feat: Add data file template download route by data category

Azher2Ali opened the pull request:
## Summary

Adds a new route in the Lyric provider to download data file templates based on data category, using the existing Lectern client. This enables submitters to retrieve pre-populated TSV/CSV templates for each submission category.


## Description of Changes
- Added new endpoint
 - Accepts a `dataCategory` as a path parameter.
  - Optional query parameter `format` to select file type (`tsv` or `csv`, default is `tsv`).
  - Uses existing Lectern client integration to fetch the corresponding dictionary schema for the category.
  - Generates a data file template with appropriate headers.
  - Sends the template file content as a download with appropriate file name and content type.

- **Exported utility function** from Lectern client to create template content by schema.

- **Updated Lyric provider** export to include the new route, making it available to services like PCGL submission-api without additional implementation.



### Special Instructions
Before running these changes, you will need to run 
`pnpm i`


## Readiness Checklist

- [x] **Self Review**
  - I have performed a self review of code
  - I have run the application locally and manually tested the feature
  - I have checked all updates to correct typos and misspellings
- [x] **Formatting**
  - Code follows the project style guide
  - Autmated code formatters (ie. Prettier) have been run
- [x] **Local Testing**
  - Successfully built all packages locally
  - Successfully ran all test suites, all unit and integration tests pass

leoraba commented on packages/data-provider/src/controllers/categoryController.ts:
```diff
@@ -1,15 +1,20 @@
+import JSZip from 'jszip';
+import dictionaryService from 'src/services/dictionaryService.js';
+
+import { createDataFileTemplate } from '@overture-stack/lectern-client';
```
This function was added in lectern-client `2.0.0-beta.5`. Make sure lectern dependency is updated

leoraba commented on packages/data-provider/src/controllers/categoryController.ts:
```diff
@@ -37,6 +42,81 @@ const controller = (dependencies: BaseDependencies) => {
+				}
+
+				const fileType = fileTypeQueryParse.data;
+				const dictionary = replaceReferences(await service.fetchDictionaryByVersion(name, version));
```
where `replaceReferences` function is imported from?

leoraba commented on packages/data-provider/src/controllers/categoryController.ts:
```diff
@@ -37,6 +42,81 @@ const controller = (dependencies: BaseDependencies) => {
 			}
 		},
+		downloadDataFileTemplates: async (
+			req: Request<{}, {}, {}, { name: string; version: string; fileType?: string }>,
```
to standard the request validation, use `validateRequest` middleware similar than other controllers.

leoraba reviewed (changes requested):
This PR doesn't compile, require some changes as commented.

leoraba commented on packages/data-provider/src/routers/categoryRouter.ts:
```diff
@@ -19,6 +19,9 @@ const router = ({
 	router.get('/', categoryController(baseDependencies).listAll);
 	router.get('/:categoryId', categoryController(baseDependencies).getDetails);
+
+	router.get('/: categoryId/dictionary', categoryController(baseDependencies).getDictionaryJson);
```
there is an empty space `: categoryId`

leoraba commented on packages/data-provider/src/controllers/categoryController.ts:
```diff
@@ -1,15 +1,20 @@
 import { NextFunction, Request, Response } from 'express';
+import JSZip from 'jszip';
+import dictionaryService from 'src/services/dictionaryService.js';
```
change this import to relative import

leoraba commented on packages/data-provider/src/controllers/categoryController.ts:
```diff
@@ -37,6 +42,73 @@ const controller = (dependencies: BaseDependencies) => {
+				}
+			},
+		),
+		getDictionaryJson: async (req: Request, res: Response, next: NextFunction) => {
```
Same as previous comment. Remove the `Request`, `Response` and `NextFunction`. params should be infer by the request validator

leoraba commented on packages/data-provider/src/controllers/dictionaryController.ts:
```diff
@@ -42,6 +49,76 @@ const controller = (dependencies: BaseDependencies) => {
 		}),
+		downloadDataFileTemplates: validateRequest(downloadDataFileTemplatesSchema, async (req, res, next) => {
+			try {
+				const { name, version, fileType } = downloadDataFileTemplatesSchema.query.parse(req.query);
+
+				const dictionary = await dictionaryService.fetchDictionaryByVersion(name, version);
+
+				if (!dictionary) {
+					throw new NotFound(`Dictionary with name "${name}" and version "${version}" not found.`);
+				}
```
I think the only required param in the request could be `categoryId` and `fileType`. 
Then, the dictionary can be fetched by using the `dictionaryService.getActiveDictionaryByCategory(categoryId)`.

leoraba commented on apps/server/swagger/dictionary-api.yml:
```diff
@@ -44,3 +44,81 @@
+        schema:
+          type: integer
+        description: The ID of the category containing the dictionary
+      - name: name
+        in: query
+        required: true
+        schema:
+          type: string
+        description: Dictionary name
+      - name: version
+        in: query
+        required: true
+        schema:
+          type: string
+        description: Dictionary version
```
`name` and `version` not required as we're using its `categoryId`

Azher2Ali commented:
> Code is looking good just small changes requested.

- Updated dictionary routes to follow /dictionary/category/:categoryId pattern
- Removed unused dictId and schemaName params from controller methods
- Simplified downloadDataFileTemplates to require only categoryId and fileType
- Refactored getDictionaryJson to return entire dictionary JSON by categoryId
- Updated Swagger documentation to reflect new route paths and request parameters
- Removed unused getOneById method from dictionaryService

leoraba commented on apps/server/swagger/dictionary-api.yml:
```diff
@@ -44,3 +44,69 @@
+        schema:
+          type: integer
+        description: The ID of the category containing the dictionary
+      - name: schemaName
+        in: query
+        required: true
+        schema:
+          type: string
+        description: Name of the schema to fetch from the dictionary
```
`schemaName` query param is required in swagger but not used in the controller code

leoraba commented on packages/data-provider/src/controllers/dictionaryController.ts:
```diff
@@ -42,6 +49,54 @@ const controller = (dependencies: BaseDependencies) => {
 		}),
+		downloadDataFileTemplates: validateRequest(downloadDataFileTemplatesSchema, async (req, res, next) => {
+			try {
+				const { fileType, categoryId } = downloadDataFileTemplatesSchema.query.parse(req.query);
```
do not need to parse `req` in this line, it has been already parsed by the `validateRequest` function

leoraba commented on packages/data-provider/src/controllers/dictionaryController.ts:
```diff
@@ -1,7 +1,14 @@
+
 import { BaseDependencies } from '../config/config.js';
 import dictionarySvc from '../services/dictionaryService.js';
+import { BadRequest, NotFound } from '../utils/errors.js';
```
Unnecessary import of the class `BadRequest` class, It is not used in this file.

leoraba reviewed (changes requested):
Added some comments regarding the request input and validation

leoraba reviewed (approved):
LGTM
